File: 0201-0300/0272-Closest Binary Search Tree Value II/0272-Closest Binary Search Tree Value II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> closestKValues(TreeNode* root, double target, int k) {
        stack<TreeNode*> pre;
        stack<TreeNode*> suc;
        while (root) {
            if (root->val <= target) {
                pre.push(root);
                root = root->right;
            }
            else {
                suc.push(root);
                root = root->left;
            }
        }
        
        vector<int> result;
        while (k--) {
            if (pre.empty()) {
                result.push_back(getNextSuccessor(suc));
            }
            else if (suc.empty()) {
                result.push_back(getNextPredecessor(pre));
            }
            else {
                if (abs(pre.top()->val - target) < abs(suc.top()->val - target)) {
                    result.push_back(getNextPredecessor(pre));
                }
                else {
                    result.push_back(getNextSuccessor(suc));
                }
            }
        }
        
        return result;
    }
    
private:
    int getNextPredecessor(stack<TreeNode*>& pre) {
        TreeNode* curr = pre.top();
        pre.pop();
        int res = curr->val;
        curr = curr->left;
        while (curr) {
            pre.push(curr);
            curr = curr->right;
        }
        
        return res;
    }
    
    int getNextSuccessor(stack<TreeNode*>& suc) {
        TreeNode* curr = suc.top();
        suc.pop();
        int res = curr->val;
        curr = curr->right;
        while (curr) {
            suc.push(curr);
            curr = curr->left;
        }
        
        return res;
    }
};

File: 0201-0300/0272-Closest Binary Search Tree Value II/0272-Closest Binary Search Tree Value II.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def closestKValues(self, root: TreeNode, target: float, k: int) -> List[int]:
        pre = []
        suc = []
        def getNextPredecessor():
            curr = pre.pop()
            val = curr.val
            curr = curr.left
            while curr:
                pre.append(curr)
                curr = curr.right
            return val
        
        def getNextSuccessor():
            curr = suc.pop()
            val = curr.val
            curr = curr.right
            while curr:
                suc.append(curr)
                curr = curr.left
            return val
        
        while root:
            if root.val <= target:
                pre.append(root)
                root = root.right
            else:
                suc.append(root)
                root = root.left
        
        result = []
        for _ in range(k):
            if not pre:
                result.append(getNextSuccessor())
            elif not suc:
                result.append(getNextPredecessor())
            else:
                if target - pre[-1].val < suc[-1].val - target:
                    result.append(getNextPredecessor())
                else:
                    result.append(getNextSuccessor())
        
        return result
